### PR TITLE
Warn if extension receiver already has member

### LINKED ISSUE #16743: Java additions can silently break overloading extension methods
I used Java 11 for a long time and I had an extension method `indent` on `String` that indents a string.
Java 12 and above introduced its own `indent` method. As a result, my existing code broke, like demonstrated in the following code.
## Compiler version

v3.2.2-RC2

## Minimized code
```Scala
extension (text: String)
  def indent: String = text

println("hello".indent)
```

## Output
Under Java 12 and above:
```scala
Playground$$$Lambda$7955/0x00000008010fc800@4e7d4166
```

## Expectation
Print `hello` or at least give a warning/error that overloading existing class methods via extension methods is not possible.

### PR DESCRIPTION
**Pull request by som-snytt**
Best effort to warn if defining extension method for a type which has a non-private member of the same name and a subsuming signature.

Excludes opaque types (for defining forwarders to a public API) and other alias types (where the type member may override an abstract type member).

Fixes #16743

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala [hunk 1]
```diff
@@ -1348,6 +1348,7 @@ object SymDenotations {
      *  inClass <-- find denot.symbol      class C { <-- symbol is here
      *
      *                   site: Subtype of both inClass and C
+     * } <-- balance the brace
```
Comment (som-snytt) [inline compiler/src/dotty/tools/dotc/core/SymDenotations.scala:1351]: braceful folly!

File: compiler/src/dotty/tools/dotc/reporting/ErrorMessageID.scala [hunk 2]
```diff
@@ -196,6 +196,7 @@ enum ErrorMessageID(val isActive: Boolean = true) extends java.lang.Enum[ErrorMe
   case AmbiguousExtensionMethodID // errorNumber 180
   case UnqualifiedCallToAnyRefMethodID // errorNumber: 181
   case NotConstantID // errorNumber: 182
+  case ExtensionNullifiedByMemberID // errorNumber: 183
```
Comment (som-snytt) [inline compiler/src/dotty/tools/dotc/reporting/ErrorMessageID.scala:199]: I repaired the conflict, but it seems to me that MiMA should be enlisted to check that `errorNumber` is fixed once committed.

**Comment by sjrd** [comment]
@jchyb ping for triage. IIUC you're issue supervisor this week?

File: compiler/src/dotty/tools/dotc/reporting/ErrorMessageID.scala (hunk 2 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/reporting/ErrorMessageID.scala:199]: I repaired it again. Happy 2024.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 3]
```diff
@@ -2596,7 +2596,14 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         linkConstructorParams(sym, tparamSyms, rhsCtx)
 
     if sym.isInlineMethod then rhsCtx.addMode(Mode.InlineableBody)
-    if sym.is(ExtensionMethod) then rhsCtx.addMode(Mode.InExtensionMethod)
+    if sym.is(ExtensionMethod) then
+      if ctx.phase.isTyper then
+        val ValDef(_, paramTpt, _) = termParamssIn(paramss1).head.head
+        if !paramTpt.symbol.denot.isAliasType && !paramTpt.symbol.denot.isOpaqueAlias then
+          val other = paramTpt.tpe.nonPrivateMember(name)
+          if other.exists && sym.denot.info.resultType.matches(other.info) then
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:2604]: There's a few situations where this check wouldn't be enough:

1. The method defined in the class is overloaded (`other.exists` will return false).

2. The parameters in the extension definition are subtypes of the one in the primary definition:
```scala
class C { def foo(x: Any) = 0 }
extension (c: C) def foo(x: Int) = 1 // uncallable
```

2. Only the first (if any) parameter list matches:
```scala
class C { def foo(x: Int)(y: String) = 0 }
extension (c: C) def foo(x: Int)(y: Int) = 1 // uncallable
```
```scala
class C { def foo: Int = 0 }
extension (c: C) def foo: String = "" // uncallable
```

4. The primary definition might have a type parameter or implicit/using clause:
```scala
class C { def foo(using Int) = 0 }
extension (c: C) def foo = 1 // uncallable
```

5. The extension definition might have a type parameter or implicit/using clause in its prefix:
```scala
class C { def foo = 0 }
extension (using Int)(c: C) def foo = 1 // uncallable
```

6. Borderline (which we may or may not want to warn about):  the extension definition might have a type parameter or implicit/using clause in its suffix:
```scala
class C { def foo = 0 }
extension (c: C) def foo(using Int) = 1 // only callable by explicitly passing `(using ...)`
```

I think this is solvable by doing something like:

```suggestion
          def firstExplicitParamTypes(tp: Type) = stripImplicit(tp.stripPoly, wildcardOnly = true).firstParamTypes
          val extFirst = firstExplicitParamTypes(sym.denot.info)
          val matching =
            paramTpt.tpe.nonPrivateMember(name)
              .filterWithPredicate: clsMeth =>
                val clsFirst = firstExplicitParamTypes(clsMeth.info)
                extFirst.lazyZip(clsFirst).forall(_ frozen_<:< _)
              .nonEmpty
```
But I haven't actually tried it.

**Review by smarter — APPROVED** [review]
Oops, looks like I completely missed this. I think this is fine as-is since it shouldn't have any false positives but I left a comment with a possible improvement to avoid false negatives.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 3 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:2604]: Thanks! I was bored and in need of stimulation. If I'm able to work through the cases, I'm sorry in advance that I will request your re-review.

Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:2604]: case 7 is case 6 but where `c.foo(using x)` typechecks with native member returning `{ def apply(using X) }`. I had to refresh my knowledge, and the reference is not current or precise (as it mentions only when m is missing, not when m is present but expression doesn't typecheck). I'll experiment now with your code suggestion.

**Comment by som-snytt** [comment]
Ha, it errored on an old line of mine. The test almost covered it: member `m: String` hides extension `m(i: Int)` because `m(42)` is `String#apply`, whereas `m: Int` does not hide.

Result type does not drive extension search, so `m: String` always hides nullary extension.

For non-nullary extension and nullary member, consider apply in result.

The existing test case `n` was only about nullary member and extension with implicit params. We need a word, "implicitary", for that.

**Comment by som-snytt** [comment]
Moved the check to refchecks, where it is more at home.

Added some test code and conditions, learned something about extension methods and something about dotty internals!

**Comment by som-snytt** [comment]
Already missed the narrow window to claim that message ID.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 4]
```diff
@@ -1048,6 +1047,38 @@ object RefChecks {
...
 
+  /** Check that an extension method is not hidden, i.e., that it is callable.
+   *
+   *  An extension method is hidden if it does not offer a parameter that is not subsumed
+   *  by the corresponding parameter of the member (or of all alternatives of an overload).
+   *
+   *  If the member has implicit parameters, then the application must supply explicit `using`
+   *  arguments to select an extension method. The reason is that if the arguments are not
+   *  supplied implicitly, typechecking fails and stops. But with explicit arguments, extensions
+   *  may be considered, and of course the extension must also accept implicit parameters.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1058]: This description is a little hard to follow, maybe rework it with examples?

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 5]
```diff
@@ -1048,6 +1047,38 @@ object RefChecks {
...
+   *
+   *  If the extension method is nullary, it is always hidden by a member of the same name.
+   */
+  def checkExtensionMethods(sym: Symbol)(using Context): Unit = if sym.is(Extension) then
+    extension (tp: Type)
+      def firstExplicitParamTypes = Applications.stripImplicit(tp.stripPoly, wildcardOnly = true).firstParamTypes
+      def hasImplicitParams = tp.stripPoly match { case mt: MethodType => mt.isImplicitMethod case _ => false }
+    val target = sym.info.firstParamTypes.head // required for extension method
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1066]: I believe this should be `firstExplicitParamTypes` to account for `extension (using f: Foo)(c: f.MyOpaque) def foo = ...`

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 6]
```diff
@@ -1048,6 +1047,38 @@ object RefChecks {
...
+    if !target.typeSymbol.denot.isAliasType && !target.typeSymbol.denot.isOpaqueAlias then
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1067]: Not warning for opaque types makes sense (though really it depends on the scope where the extension is defined), but what's special about aliases?

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 7]
```diff
@@ -1048,6 +1047,38 @@ object RefChecks {
...
+      val methTp = sym.denot.info.resultType
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1068]: That also looks like it will do the wrong thing for extensions starting with a using clause (also `sym.denot.info` is equivalent to `sym.info`), e.g.:
```scala
class C { def foo(x: Int): Int = 0 }
extension (using Int)(c: C) def foo(x: Int): Int = 1
```

File: compiler/src/dotty/tools/dotc/reporting/messages.scala [hunk 8]
```diff
@@ -2429,6 +2429,15 @@ class SynchronizedCallOnBoxedClass(stat: tpd.Tree)(using Context)
...
 }
 
+class ExtensionNullifiedByMember(method: Symbol, target: Symbol)(using Context)
+  extends Message(ExtensionNullifiedByMemberID):
+  def kind = MessageKind.PotentialIssue
+  def msg(using Context) = i"Suspicious extension ${hl(method.name.toString)} is already a member of ${hl(target.name.toString)}"
+  def explain(using Context) =
+    i"""Extension method ${hl(method.name.toString)} will never be selected
+       |because ${hl(target.name.toString)} already has a member with the same name.
+       |It can be called as a regular method, but should probably be defined that way."""
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/reporting/messages.scala:2439]: "Suspicious" is a bit vague, but the explanation is good, so I would just replace the content of `msg` by `explain` here.

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala (hunk 6 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1067]: opaque also just works as expected, so I removed the line

Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: or maybe there was a reason my test class was `IArray`.

Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: so I added a more explicit test class called `Tree`. and added back the exclusion for aliases.

In future, it could check if the type member doesn't override and maybe warn in that case.

**Comment by som-snytt** [comment]
Followed up on previous review.

File: compiler/src/dotty/tools/dotc/reporting/messages.scala [hunk 9]
```diff
@@ -2451,6 +2451,17 @@ class SynchronizedCallOnBoxedClass(stat: tpd.Tree)(using Context)
         |you intended."""
 }
 
+class ExtensionNullifiedByMember(method: Symbol, target: Symbol)(using Context)
+  extends Message(ExtensionNullifiedByMemberID):
+  def kind = MessageKind.PotentialIssue
+  def msg(using Context) =
+    i"""Extension method ${hl(method.name.toString)} will never be selected
+       |because ${hl(target.name.toString)} already has a member with the same name."""
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/reporting/messages.scala:2459]: Maybe we can be a bit more precise without getting into too many details:
```suggestion
       |because ${hl(target.name.toString)} already has a member with the same name and compatible parameter types."""
```

**Review by smarter — APPROVED** [review]
Otherwise LGTM, thanks for your work!

**Comment by som-snytt** [comment]
Last chance for this PR to grab an error ID before other PRs overtake it. https://github.com/scala/scala3/pull/20143

**Comment by smarter** [comment]
:1st_place_medal:

**Comment by soronpo** [comment]
This PR creates false positives under 3.5.0-RC7

**Comment by soronpo** [comment]
I'm sorry, I phrased it incorrectly. These are not false positives, but they create warnings for definitions I use as workaround for lack of imports in an extension method block. 

Here is an example:
```scala
class Foo(val someField: Int)

extension (foo: Foo)
  def someField = foo.someField
  def bar1 = 1 + someField
  def bar2 = 2 + someField
  def bar3 = 3 + someField
```
Since I cannot apply `import foo.someField` commonly in the extension method block, this concept has been a useful hack.
@som-snytt What do you think about excluding the specific pattern of `def <fieldName> = extended.<fieldName>` from the warning?
